chore(docs): Core concepts page in getting-started (#3401)
AztecProtocol/aztec-packages: docs/docs/dev_docs/getting_started/core-concepts.md

* The PXE and the Public VM cannot directly communicate with each other
* Private transactions in the PXE are executed first, followed by public transactions

You can call a public function from a private function by using this:

#include_code call_public_function yarn-project/noir-contracts/src/contracts/card_game_contract/src/main.nr rust

You cannot call a private function from a public function, but you can use a
slow updates tree to read and write to public state from a private function. 

### Data types

Private state works with UTXOs, or what we call notes. To keep things private, everything is stored in an append-only UTXO tree, and a nullifier is created when notes are spent.

Public state works similarly to other chains like Ethereum, behaving more like a public ledger. 

Working with private state is like creating commitments and nullifiers, whereas working with public state is like directly updating state.

We have abstractions for working with private state so you don't have to worry about these commitments and nullifiers. However, it is important to understand that the types and libraries you use will be different when working with private state and public state.

> For example, let's say you're trying to work with an integer. We have a library called `EasyPrivateUint` that acts like an int but in the background is actually updating notes in private state. For the public side, we instead have something called `SafeU120`. You cannot use EasyPrivateUint in a public environment, and you cannot use SafeU120 in a private environment.

Reviewer: ```suggestion
For example, let's say you're trying to work with an integer. We have a library called `EasyPrivateUint` that acts like an integer but in the background is actually updating notes in private state. For the public side, we instead have something called `SafeU120`. You cannot use EasyPrivateUint in a public environment, and you cannot use SafeU120 in a private environment.
```

# Storage

Currently, when writing Aztec.nr smart contracts, you will need to define two things when initiating storage:

1. The storage struct, ie what you are storing and their types
2. A storage `impl` block with `init` function that will be called when you use the storage in a function

The storage struct looks like this:

#include_code storage_struct yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

The storage impl block looks like this:

#include_code storage_init yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

The `init` function must declare the storage struct with an instantiation defining how variables are accessed and manipulated. Each variable must be given a storage slot, which can be anything except 0.

The impl block is likely to be abstracted away at a later date.

> 

Reviewer: Should we a link to [this](https://docs.aztec.network/dev_docs/contracts/syntax/storage/main) page for further reading?

# Portals

Aztec allows you to interact with Ethereum privately - ie no-one knows where the transaction is coming from, just that it is coming from somewhere on Aztec.

This is achieved through portals - these are smart contracts written in Solidity that are related to the Ethereum smart contract you want to interact with.

A portal can be associated with multiple Aztec contracts, but an Aztec contract can only be associated with one portal. 

> 

Reviewer: Should we add a link to [this](https://docs.aztec.network/dev_docs/contracts/portals/main) page for further reading?

# Account Abstraction

Every account in Aztec is a smart contract. This allows implementing different schemes for transaction signing, nonce management, and fee payments.

You can write your own account contract to define the rules by which user transactions are authorized and paid for, as well as how user keys are managed.

> 

Reviewer: Should we add a link to this page? https://docs.aztec.network/concepts/foundation/accounts/main

# Noir Language

> Aztec smart contracts are written in a framework on top of Noir, the zero-knowledge domain-specific language developed specifically for Aztec. It is similar to Rust. Outside of Aztec, Noir is used for writing circuits that can be verified in Solidity.

Reviewer: ```suggestion
Aztec smart contracts are written in a framework on top of Noir, the zero-knowledge domain-specific language developed specifically for Aztec. It's syntax is similar to Rust. Outside of Aztec, Noir is used for writing circuits that can be verified in Solidity.
```

> You do not need to understand Noir to write Aztec contracts. However, it may be useful to reference the [Noir docs](https://noir-lang.org) when you start writing more advanced contracts.

Reviewer: ```suggestion
A cursory understanding of Noir is sufficient for writing Aztec contracts. The [Noir docs](https://noir-lang.org) will be a helpful reference when you start writing more advanced contracts.
```
